File: backend/src/auto_stock_trading/domain/market_data/etf.py
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from datetime import date, datetime
    from decimal import Decimal

    from auto_stock_trading.domain.market_data.models import RawBrokerResponse


@dataclass(frozen=True, slots=True)
class EtfProfile:
    symbol: str
    isin: str
    name: str
    source: str
    received_at: datetime


@dataclass(frozen=True, slots=True)
class VersionedEtfProfile:
    symbol: str
    isin: str
    name: str
    source: str
    received_at: datetime
    version: int
    valid_from: datetime
    superseded_at: datetime | None


@dataclass(frozen=True, slots=True)
class EtfMasterBundle:
    profiles: tuple[EtfProfile, ...]
    raw: RawBrokerResponse
    collected_at: datetime


@dataclass(frozen=True, slots=True)
class EtfNavSnapshot:
    symbol: str
    price: Decimal
    change_percent: Decimal
    volume: int
    previous_volume: int
    nav: Decimal
    divergence_rate: Decimal
    tracking_error: Decimal
    tracking_multiple: Decimal
    net_asset_total: int
    listed_shares: int
    manager: str
    index_name: str
    listing_date: date | None
    currency: str
    source: str
    as_of: datetime
    received_at: datetime


@dataclass(frozen=True, slots=True)
class EtfNavObservation:
    snapshot: EtfNavSnapshot
    raw: RawBrokerResponse


@dataclass(frozen=True, slots=True)
class EtfListing:
    profile: VersionedEtfProfile
    snapshot: EtfNavSnapshot | None

File: frontend/src/lib/etf.ts
import { z } from "zod";

const decimal = z.string().regex(/^-?\d+(\.\d+)?([Ee][+-]?\d+)?$/);
const nullableDecimal = decimal.nullable();
const isoDate = z.iso.date();
const isoDateTime = z.iso.datetime();

const etfSnapshotSchema = z.strictObject({
  as_of: isoDateTime,
  change_percent: decimal,
  currency: z.string().min(1),
  divergence_rate: decimal,
  index_name: z.string().min(1),
  listed_shares: z.number().int(),
  listing_date: isoDate.nullable(),
  manager: z.string().min(1),
  nav: decimal,
  net_asset_total: z.number().int(),
  previous_volume: z.number().int(),
  price: decimal,
  received_at: isoDateTime,
  tracking_error: decimal,
  tracking_multiple: decimal,
  volume: z.number().int(),
});

const etfListingSchema = z.strictObject({
  isin: z.string().min(1),
  name: z.string().min(1),
  snapshot: etfSnapshotSchema.nullable(),
  symbol: z.string().min(1),
});

const etfsSchema = z.strictObject({
  etfs: z.array(etfListingSchema).readonly(),
  master_source: z.string().min(1),
  net_asset_unit: z.literal("hundred_million_krw"),
  source: z.string().min(1),
});

const distributionYieldSchema = z.strictObject({
  distribution_count: z.number().int(),
  distribution_total: nullableDecimal,
  formula: z.string().min(1),
  unavailable_reason: z
    .enum(["MISSING_SNAPSHOT", "MISSING_DISTRIBUTIONS", "ZERO_PRICE"])
    .nullable(),
  value: nullableDecimal,
  window_end: isoDate.nullable(),
  window_start: isoDate.nullable(),
});

const etfDetailSchema = z.strictObject({
  distribution_yield: distributionYieldSchema,
  isin: z.string().min(1),
  master_source: z.string().min(1),
  name: z.string().min(1),
  net_asset_unit: z.literal("hundred_million_krw"),
  snapshot: etfSnapshotSchema.nullable(),
  source: z.string().min(1),
  symbol: z.string().min(1),
});

export type EtfSnapshot = z.infer<typeof etfSnapshotSchema>;
export type EtfListing = z.infer<typeof etfListingSchema>;
export type Etfs = z.infer<typeof etfsSchema>;
export type EtfDetail = z.infer<typeof etfDetailSchema>;

export const parseEtfs = (input: unknown): Etfs => etfsSchema.parse(input);
export const parseEtfDetail = (input: unknown): EtfDetail => etfDetailSchema.parse(input);
